File: examples/typescript/virtual_x360_pad.ts
import { ViiperClient, ViiperDevice, Xbox360, Types } from "viiperclient";

const { Xbox360Input, Button } = Xbox360;

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

// Minimal example: ensure a bus, create an xbox360 device, stream inputs, read rumble, clean up on exit.
async function main() {
  if (process.argv.length < 3) {
    console.log("Usage: node virtual_x360_pad.js <api_addr>");
    console.log("Example: node virtual_x360_pad.js localhost:3242");
    process.exit(1);
  }

  const addr = process.argv[2];
  const [host, portStr] = addr.split(':');
  const port = portStr ? parseInt(portStr, 10) : 3242;
  const client = new ViiperClient(host, port);

  // Find or create a bus
  const busesResp = await client.buslist();
  let busID: number;
  let createdBus = false;
  
  if (busesResp.buses.length === 0) {
    let createErr: any;
    busID = 0;
    for (let tryBus = 1; tryBus <= 100; tryBus++) {
      try {
        const r = await client.buscreate(tryBus);
        busID = r.busId;
        createdBus = true;
        break;
      } catch (err) {
        createErr = err;
      }
    }
    if (busID === 0) {
      console.error(`BusCreate failed: ${createErr}`);
      process.exit(1);
    }
    console.log(`Created bus ${busID}`);
  } else {
    busID = Math.min(...busesResp.buses);
    console.log(`Using existing bus ${busID}`);
  }

  // Add device and connect to stream in one call
  let dev: ViiperDevice;
  let deviceDevId: string;
  try {
    const req: Types.DeviceCreateRequest = { type: "xbox360" };
    const { device, response: addResp } = await client.addDeviceAndConnect(busID, req);
    dev = device;
    deviceDevId = addResp.devId;
    console.log(`Created and connected to device ${deviceDevId} on bus ${busID}`);
  } catch (err) {
    console.error(`AddDeviceAndConnect error: ${err}`);
    if (createdBus) {
      await client.busremove(busID).catch(() => {});
    }
    process.exit(1);
  }

  // Cleanup function
  const cleanup = async () => {
    try {
      dev.close();
      await client.busdeviceremove(busID, deviceDevId);
      console.log(`Removed device ${deviceDevId}`);
    } catch (err) {
      console.error(`DeviceRemove error: ${err}`);
    }
    if (createdBus) {
      try {
        await client.busremove(busID);
        console.log(`Removed bus ${busID}`);
      } catch (err) {
        console.error(`BusRemove error: ${err}`);
      }
    }
  };

  // Start event-driven rumble reading (2 bytes per rumble state)
  dev.on("output", (buf: Buffer) => {
    if (buf.length >= 2) {
      const leftMotor = buf.readUInt8(0);
      const rightMotor = buf.readUInt8(1);
      console.log(`← Rumble: Left=${leftMotor}, Right=${rightMotor}`);
    }
  });

  dev.on("error", async (err: Error) => {
    console.error(`Stream error: ${err}`);
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(1);
  });

  dev.on("end", async () => {
    console.log("Stream ended by server");
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(0);
  });

  // Handle signals for graceful shutdown
  process.on("SIGINT", async () => {
    console.log("Signal received, stopping…");
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(0);
  });
  process.on("SIGTERM", async () => {
    console.log("Signal received, stopping…");
    running = false;
    clearInterval(interval);
    await cleanup();
    process.exit(0);
  });

  // Send controller inputs at 60fps (16ms intervals)
  let frame = 0;
  let running = true;
  const interval = setInterval(async () => {
    if (!running) return;
    
    try {
      frame++;
      let buttons = 0;
      switch (Math.floor((frame / 60) % 4)) {
        case 0:
          buttons = Button.A;
          break;
        case 1:
          buttons = Button.B;
          break;
        case 2:
          buttons = Button.X;
          break;
        default:
          buttons = Button.Y;
          break;
      }
      
      const state = new Xbox360Input({
        Buttons: buttons,
        Lt: (frame * 2) % 256,
        Rt: (frame * 3) % 256,
        Lx: Math.floor(20000.0 * 0.7071),
        Ly: Math.floor(20000.0 * 0.7071),
        Rx: 0,
        Ry: 0,
      });
      
      await dev.send(state);
      
      if (frame % 60 === 0) {
        console.log(`→ Sent input (frame ${frame}): buttons=0x${state.Buttons.toString(16).padStart(4, "0")}, LT=${state.Lt}, RT=${state.Rt}`);
      }
    } catch (err) {
      console.error(`Write error: ${err}`);
      running = false;
      clearInterval(interval);
      await cleanup();
      process.exit(1);
    }
  }, 16);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/cpp/virtual_x360_pad.cpp
#define VIIPER_JSON_INCLUDE <nlohmann/json.hpp>
#define VIIPER_JSON_NAMESPACE nlohmann
#define VIIPER_JSON_TYPE json

#include <viiper/viiper.hpp>
#include <iostream>
#include <thread>
#include <chrono>
#include <csignal>
#include <atomic>
#include <cmath>

std::atomic<bool> running{true};

void signal_handler(int) {
    running = false;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <api_addr>\n";
        std::cerr << "Example: " << argv[0] << " localhost:3242\n";
        return 1;
    }

    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    const std::string addr = argv[1];
    const auto colon_pos = addr.find(':');
    const std::string host = addr.substr(0, colon_pos);
    const std::uint16_t port = colon_pos != std::string::npos
        ? static_cast<std::uint16_t>(std::stoul(addr.substr(colon_pos + 1)))
        : 3242;

    viiper::ViiperClient client(host, port);

    // Find or create a bus
    std::uint32_t bus_id;
    bool created_bus = false;

    auto buses_result = client.buslist();
    if (buses_result.is_error()) {
        std::cerr << "BusList error: " << buses_result.error().to_string() << "\n";
        return 1;
    }

    if (buses_result.value().buses.empty()) {
        auto create_result = client.buscreate(std::nullopt);
        if (create_result.is_error()) {
            std::cerr << "BusCreate failed: " << create_result.error().to_string() << "\n";
            return 1;
        }
        bus_id = create_result.value().busid;
        created_bus = true;
        std::cout << "Created bus " << bus_id << "\n";
    } else {
        bus_id = buses_result.value().buses[0];
        std::cout << "Using existing bus " << bus_id << "\n";
    }

    // Add device
    auto device_result = client.busdeviceadd(bus_id, {.type = "xbox360"});
    if (device_result.is_error()) {
        std::cerr << "AddDevice error: " << device_result.error().to_string() << "\n";
        if (created_bus) {
            client.busremove(bus_id);
        }
        return 1;
    }
    auto device_info = std::move(device_result.value());

    // Connect to device stream
    auto stream_result = client.connectDevice(device_info.busid, device_info.devid);
    if (stream_result.is_error()) {
        std::cerr << "ConnectDevice error: " << stream_result.error().to_string() << "\n";
        client.busdeviceremove(device_info.busid, device_info.devid);
        if (created_bus) {
            client.busremove(bus_id);
        }
        return 1;
    }
    auto stream = std::move(stream_result.value());

    std::cout << "Created and connected to device " << device_info.devid
              << " on bus " << device_info.busid << "\n";

    stream->on_disconnect([]() {
        std::cerr << "Device disconnected by server\n";
        std::exit(0);
    });

    stream->on_output(viiper::xbox360::OUTPUT_SIZE, [](const std::uint8_t* data, std::size_t len) {
        if (len < viiper::xbox360::OUTPUT_SIZE) return;
        auto result = viiper::xbox360::Output::from_bytes(data, len);
        if (result.is_error()) return;
        auto& rumble = result.value();
        std::cout << "← Rumble: Left=" << static_cast<int>(rumble.left)
                  << ", Right=" << static_cast<int>(rumble.right) << "\n";
    });

    // Send controller inputs at 60fps (16ms intervals)
    std::uint64_t frame = 0;

    while (running && stream->is_connected()) {
        ++frame;

        std::uint16_t buttons;
        switch ((frame / 60) % 4) {
            case 0: buttons = viiper::xbox360::ButtonA; break;
            case 1: buttons = viiper::xbox360::ButtonB; break;
            case 2: buttons = viiper::xbox360::ButtonX; break;
            default: buttons = viiper::xbox360::ButtonY; break;
        }

        viiper::xbox360::Input state = {
            .buttons = buttons,
            .lt = static_cast<std::uint8_t>((frame * 2) % 256),
            .rt = static_cast<std::uint8_t>((frame * 3) % 256),
            .lx = static_cast<std::int16_t>(20000.0 * 0.7071),
            .ly = static_cast<std::int16_t>(20000.0 * 0.7071),
            .rx = 0,
            .ry = 0,
        };

        auto send_result = stream->send(state);
        if (send_result.is_error()) {
            std::cerr << "Write error: " << send_result.error().to_string() << "\n";
            break;
        }

        if (frame % 60 == 0) {
            std::cout << "→ Sent input (frame " << frame << "): buttons=0x"
                      << std::hex << state.buttons << std::dec
                      << ", LT=" << static_cast<int>(state.lt)
                      << ", RT=" << static_cast<int>(state.rt) << "\n";
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(16));
    }

    // Cleanup
    stream->stop();
    client.busdeviceremove(device_info.busid, device_info.devid);
    if (created_bus) {
        client.busremove(bus_id);
    }

    return 0;
}
